Feature/phrase spec errors

mgrbyte opened the pull request:
Provides:
 - human readable error message when a validation error occurs (HTTP 400 responses).
 - display "primary ID" (i.e WB* ids)  upon db conflict (.eg "$name already stored against $wbid)
 
cases tested (manually)
 -  gene create
 -  var create
 -  strain create

Error messages  `clojure.spec` failures are translated to human readable messages using the [phrase][1] library. 
The top level response body should contain `{"message": "...", ...}`
Species specific name validation occurs later in the processing phase, and are not implemented using `clojure.spec`,  and return a more verbose form,
containing specific error messages for each field that failed name validation.
```json
{"message": "<top level message here>",
 "type": "...",
 "errors": [{"message": "...."}, ...]}}
```


[1]: https://github.com/alexanderkiel/phrase

Commit:
Unqualified enum values in result map for creation events.

Commit:
Depend on phrase.

Commit:
style.

Commit:
Display "primary id" of entity upon db conflict.

Commit:
Provides human messages upon validation error.

Fixes #32.

sibyl229 commented:
Hi @mgrbyte can I confirm with you that the format of the json error response hasn't changed? Only the text of message field itself? The reason I'd like to check is that the error responses for all requests are handled the same way.

Commit:
show multiple error messages from validation error

sibyl229 commented:
I pushed the UI changes for #287 to this branch.

I noticed that when creating a gene, if the species is not filled in, the `errors` field seems to return an object instead of an array, which caused the UI to break, as it expects an array.

mgrbyte commented:
> I pushed the UI changes for #287 to this branch.
> 
> I noticed that when creating a gene, if the species is not filled in, the `errors` field seems to return an object instead of an array, which caused the UI to break, as it expects an array.

Good catch! That `errors` field is superfluous in this case, will remove it from the response in this case.

Generally the 400 error responses can take in two  forms:
1. cgc/sequence name errors (as shown in the OP in this issue)
2. any other field valdiation error (biotype, species, et al)
{"message": "....", ...}

(The UI seems to be handling 2 already).

Commit:
Do not output a map for species error.

Be consistent with how all other non-name validation errors are
reported. Additionally, the value will always be nil/empty, and is
uninformative.

sibyl229 commented:
@mgrbyte thanks, it's working now. Please feel free to merge.
